File: 01.06.字符串压缩/01.06.go
func compressString(S string) string {
    sBytes := []byte(S)
    res := make([]byte, 0, len(sBytes))
    i := 0
    for i < len(sBytes) {
        j := i
        for j < len(sBytes) && S[i] == S[j] {
            j++
        }
        res = append(res, sBytes[i])
        res = append(res, []byte(strconv.Itoa(j-i))...)
        i = j
    }

    if len(res) >= len(sBytes) {
        return S
    }
    return string(res)
}

File: 01.06.字符串压缩/01.06.cpp
class Solution {
public:
    string compressString(string S) {
        int i = 0;
        string res;
        while (i < S.size()) {
            int j = i;
            while (j < S.size() && S[i] == S[j]) {
                ++j;
            }
            res += S[i];
            res += to_string(j-i);
            i = j;
        }

        if(res.size() >= S.size()) {
            return S;
        }
        return res;
    }
};